fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -126,85 +131,69 @@ def reduce(self) -> None:
-                for p in params:
-                    sz = p.numel()
-                    buffer[offset : offset + sz].copy_(p.grad.data.view(-1))  # type: ignore
-                    offset += sz
-            else:
-                # we only have a single grad to reduce
-                buffer = params[0].grad.data  # type: ignore
-
-            # Reduce
-            buffer.div_(self.world_size)  # type: ignore
-            dist.reduce(tensor=buffer, dst=params_rank, group=self.process_group)  # type: ignore
-
-            # Copy reduced grads back into their original place, or free corresponding memory
-            if params_rank == self.rank:
-                offset = 0
-                for p in params:
-                    sz = p.numel()
-                    p.grad.data.copy_(buffer[offset : offset + sz].view_as(p))  # type: ignore
-                    offset += sz
+        if not self.need_reduction or self.accumulate_grads:
+            return
+
+        self.need_reduction = False
+
+        with torch.no_grad():
+            for device, per_device in self.sharded_optimizer.per_device_params.items():
+                self._reduce_grads_task(
+                    self._reduce_buffers[device],
+                    per_device,
+                    group=self.process_group,
+                    self_rank=self.rank,
+                    world_size=self.world_size,
+                )
+
+    @staticmethod
+    def _reduce_grads_task(
+        buffer: torch.Tensor, params: List[List[Parameter]], group: Any, self_rank: int, world_size: int
+    ) -> None:
+        """Helper to reduce a list of params. The params are sorted by size, smallest first, which allows for
+        an opportunistic bucketing.
+
+        NOTE: All param gradients are assumed to exist"""
+
+        buffer_size = buffer.numel()
+
+        for rank, rank_params in enumerate(params):
+            if len(rank_params) == 0:
+                continue
+
+            restore_require_grad = []
+            requests = []
+            global_rank = OSS.get_global_rank(group, rank)
+
+            # Copy small gradients into per-GPU buffers and then async reduce
+            i_p = OSS._bucket(buffer, rank_params, copy=(global_rank == self_rank), gradients=True)
+
+            if i_p > 0:
+                buffer.div_(world_size)  # type: ignore
+                requests.append(dist.reduce(tensor=buffer, dst=global_rank, group=group, async_op=True))  # type: ignore
+
+            # Directly reduce the next grads
+            for param in rank_params[i_p:]:
+                # NOTE: workaround Gloo / leaf variable requiring grad modified in place
+                if param.requires_grad:
+                    restore_require_grad.append(param)
+                    param.requires_grad = False
+
+                param.div_(world_size)  # type: ignore
+                requests.append(dist.reduce(tensor=param, dst=global_rank, group=group, async_op=True))  # type: ignore
```

Comment:
Why are we reducing the params? Shouldn't we be reducing the grads?

Reply:
Can we add a test to that catches this.

Reply:
wrong edit, I was testing OSS only basically I overlooked that, thanks for catching this